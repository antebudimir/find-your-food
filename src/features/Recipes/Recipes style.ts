import styled from 'styled-components';

export const RecipesContainer = styled.section`
	display: grid;
	place-items: center;

	@media screen and (min-width: 700px) {
		grid-template-columns: repeat(2, 1fr);
		gap: ${(props) => props.theme.spacing.tertiary};
	}

	@media screen and (min-width: 1024px) {
		grid-template-columns: repeat(3, 1fr);
	}
`;
