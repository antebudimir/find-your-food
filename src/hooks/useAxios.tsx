import { useState, useEffect } from 'react';
import axios from 'axios';

export interface RecipeProps {
	recipe: {
		label: string;
		images: ImagesProps;
		mealType: string;
		ingredientLines: string[];
		totalTime: string;
		uri: string;
	};
}

export interface ImagesProps {
	REGULAR: ImageSource;
}

export interface ImageSource {
	url: string;
}

const useAxios = (endpoint: string) => {
	const [data, setData] = useState<RecipeProps[] | []>([]),
		[error, setError] = useState(''),
		[loading, setLoading] = useState(true);

	useEffect(() => {
		const controller = new AbortController();

		(async function fetchData() {
			try {
				const { data } = await axios(endpoint, {
					signal: controller.signal,
				});
				const response = data.hits as RecipeProps[];
				setData(response);
			} catch {
				setError('An error occured.');
				setData([]);
			} finally {
				setLoading(false);
			}
		})();

		const cleanUp = () => {
			controller.abort();
		};

		return cleanUp;
	}, [endpoint]);

	return { data, error, loading };
};

export default useAxios;
