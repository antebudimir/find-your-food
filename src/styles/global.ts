import { createGlobalStyle } from 'styled-components';
import RalewayRegular from 'fonts/Raleway-Regular.woff';
import RalewayBold from 'fonts/Raleway-Bold.woff';
import RalewayExtraBold from 'fonts/Raleway-ExtraBold.woff';

export const GlobalStyles = createGlobalStyle`
@font-face {
	font-family: 'Raleway Regular';
	src: url(${RalewayRegular}) format('woff');
	font-weight: normal;
	font-style: normal;
	font-display: swap;
}

@font-face {
	font-family: 'Raleway Bold';
	src: url(${RalewayBold}) format('woff');
	font-weight: bold;
	font-style: normal;
	font-display: swap;
}


@font-face {
    font-family: 'Raleway Extra Bold';
    src: url(${RalewayExtraBold}) format('woff');
    font-weight: bold;
    font-style: normal;
    font-display: swap;
}

* {
	outline: none;
	box-sizing: border-box;
	margin: 0;
	padding: 0;
	scroll-behavior: smooth;
}

html {
    font-family: 'Raleway Regular';
	word-wrap: break-word;
}

button,
input {
	border: none;
	display: block;
}

input,
select,
textarea,
button {
	color: inherit;
	font-size: inherit;
}

input,
select,
textarea {
	font-family: inherit;
}

h1 {
    font-family: 'Raleway Extra Bold';
}

h2,
button,
a {
	font-family: 'Raleway Bold';
}

ul,
li {
	list-style: none;
}

a {
	text-decoration: none;
}

img {
	display: block;
	max-width: 100%;
}
`;
