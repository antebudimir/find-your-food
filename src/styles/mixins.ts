import { css } from 'styled-components';

interface FlexLayoutProps {
	direction?: string;
	wrap?: string;
}

interface FlexAlignmentProps {
	justify?: string;
	align?: string;
}

export const FlexLayout = ({ direction, wrap }: FlexLayoutProps) => css`
	display: flex;
	flex-flow: ${direction} ${wrap};
`;

export const FlexAlignment = ({ justify, align }: FlexAlignmentProps) => css`
	justify-content: ${justify};
	align-items: ${align};
`;

export const hiddenVisually = css`
	position: absolute;
	clip: rect(0 0 0 0);
	clip-path: inset(50%);
	overflow: hidden;
	width: 1px;
	height: 1px;
	white-space: nowrap;
`;

export const sticky = css`
	display: flex;
	flex-flow: column nowrap;
	min-height: 89.6vh;
`;
