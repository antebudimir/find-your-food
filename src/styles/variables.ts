export const variables = {
	borderRadius: {
		primary: '0.125rem',
	},
	outline: {
		primary: '1px',
	},
	shadow: {
		primary: '#3b3f20 0px 2px 5px 0px',
		secondary: '#d45464 0px 2px 5px 0px',
		tertiary: '5px 5px 20px rgba(0, 0, 0, 0.7)',
		quaternary: '5px 5px 5px rgba(0, 0, 0, 0.7)',
	},
	sizing: {
		inputDesktopWidth: '28rem',
		summaryLinkWidth: '18.9rem', // coresponds to the width of the image
		summaryContainerWidth: '28rem',
		mainContainerWidth: '60rem',
	},
	spacing: {
		primary: '1rem',
		secondary: '0.5rem 1rem',
		tertiary: '2rem',
		quaternary: '1.5rem',
	},
	color: {
		primary: '#d45464',
		primaryLighter: '#ff6478',
		secondary: ' #ffffff',
		tertiary: '#3b3f20',
	},
	fontSize: {
		primary: '1rem',
		secondary: '1.3rem',
		tertiary: '1.5rem',
		quaternary: '2rem',
		quinary: '120%',
	},
	timing: {
		primary: '200ms',
	},
	lineHeight: {
		primary: '1.5',
		secondary: '2',
	},
};
