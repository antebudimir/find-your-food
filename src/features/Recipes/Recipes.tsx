import { RecipesContainer } from './Recipes style';
import RecipeSummary from 'features/RecipeSummary/RecipeSummary';
import Message from 'components/Elements/Message/Message';
import Spinner from 'components/Elements/Spinner/Spinner';
import useAxios from 'hooks/useAxios';

export interface SearchProps {
	searchTerm?: string;
}

const APP_ID = '548667f6';
const APP_KEY = '2f9fdc0a12e13af8ed2b0ba5c5deb2ef';

const Recipes = ({ searchTerm }: SearchProps) => {
	const endpoint = `https://api.edamam.com/api/recipes/v2?type=public&q=${searchTerm}&app_id=${APP_ID}&app_key=${APP_KEY}`,
		{ data, error, loading } = useAxios(endpoint);

	const renderRecipes = data.map((recipe) => {
		const singleRecipe = recipe.recipe,
			uniqueKey = singleRecipe.uri.slice(-32),
			{ label, images, mealType, ingredientLines, totalTime } = singleRecipe;

		return (
			<RecipeSummary
				key={uniqueKey}
				label={label}
				image={images.REGULAR.url}
				mealType={mealType}
				ingredientLines={ingredientLines}
				totalTime={totalTime}
			/>
		);
	});

	return (
		<RecipesContainer>
			{loading && <Spinner />}

			{error && <Message warning messageText={error} />}

			{data.length > 0 && renderRecipes}
		</RecipesContainer>
	);
};

export default Recipes;
