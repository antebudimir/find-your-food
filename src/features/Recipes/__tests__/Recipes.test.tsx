import { ThemeProvider } from 'styled-components';
import { variables } from 'styles/variables';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import Recipes, { SearchProps } from '../Recipes';
import useAxios, { RecipeProps } from 'hooks/useAxios';

jest.mock('hooks/useAxios');

const mockResponse: RecipeProps[] = [
	{
		recipe: {
			label: 'Test label 1',
			images: {
				REGULAR: {
					url: 'image url',
				},
			},
			mealType: 'breakfast',
			ingredientLines: ['Ingredient 1', 'Ingredient 2', 'Ingredient 3'],
			totalTime: '20',
			uri: 'http://www.edamam.com/ontologies/edamam.owl#recipe_db742742099859a8053b992dd3c1f452',
		},
	},
	{
		recipe: {
			label: 'Test label 2',
			images: {
				REGULAR: {
					url: 'image url 2',
				},
			},
			mealType: 'lunch',
			ingredientLines: ['Ingredient 1', 'Ingredient 2', 'Ingredient 3'],
			totalTime: '40',
			uri: 'http://www.edamam.com/ontologies/edamam.owl#recipe_b958fb2af215767c51d7545e2f86a722',
		},
	},
];

const props: SearchProps = {
	searchTerm: 'test',
};

const renderRecipes = (props: SearchProps) => {
	const { asFragment } = render(
		<ThemeProvider theme={variables}>
			<Recipes {...props} />
		</ThemeProvider>,
	);

	return { fragment: asFragment() };
};

describe('<Recipes /> component', () => {
	test('should fetch and render recipe titles', async () => {
		jest
			.mocked(useAxios)
			.mockReturnValue({ data: mockResponse, loading: false, error: '' });

		renderRecipes(props);

		const titleOne = await screen.findByTestId('Test label 1');
		expect(titleOne).toHaveTextContent('Test label 1');
		const titleTwo = await screen.findByTestId('Test label 2');
		expect(titleTwo).toHaveTextContent('Test label 2');
	});

	test('if the Component matches snapshot', () => {
		jest
			.mocked(useAxios)
			.mockReturnValue({ data: mockResponse, loading: false, error: '' });

		const { fragment } = renderRecipes(props);

		expect(fragment).toMatchSnapshot();
	});
});
