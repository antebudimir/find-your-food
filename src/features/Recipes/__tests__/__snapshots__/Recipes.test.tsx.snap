// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipes /> component if the Component matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="sc-bczRLJ eHfXVe"
  >
    <section
      class="sc-jSMfEi fhsqYC"
    >
      <h2
        class="sc-gKXOVf iccFfb"
        data-testid="Test label 1"
      >
        Test label 1
      </h2>
      <img
        alt="Test label 1"
        class="sc-iBkjds jBSSSV"
        src="image url"
      />
      <p
        class="sc-gsnTZi FuqXk"
        data-testid="recipe-line"
      >
        <span
          class="sc-dkzDqf IGaFb"
        >
          Meal type:
        </span>
         breakfast
      </p>
      <ul
        class="sc-hKMtZM bcFPAw"
        data-testid="recipe-list"
      >
        <span
          class="sc-eCYdqJ eqqeCB"
        >
          Ingredients:
        </span>
         
        <li>
          Ingredient 1
        </li>
        <li>
          Ingredient 2
        </li>
        <li>
          Ingredient 3
        </li>
      </ul>
      <p
        class="sc-gsnTZi FuqXk"
        data-testid="recipe-line"
      >
        <span
          class="sc-dkzDqf IGaFb"
        >
          Total time:
        </span>
         20min
      </p>
    </section>
    <section
      class="sc-jSMfEi fhsqYC"
    >
      <h2
        class="sc-gKXOVf iccFfb"
        data-testid="Test label 2"
      >
        Test label 2
      </h2>
      <img
        alt="Test label 2"
        class="sc-iBkjds jBSSSV"
        src="image url 2"
      />
      <p
        class="sc-gsnTZi FuqXk"
        data-testid="recipe-line"
      >
        <span
          class="sc-dkzDqf IGaFb"
        >
          Meal type:
        </span>
         lunch
      </p>
      <ul
        class="sc-hKMtZM bcFPAw"
        data-testid="recipe-list"
      >
        <span
          class="sc-eCYdqJ eqqeCB"
        >
          Ingredients:
        </span>
         
        <li>
          Ingredient 1
        </li>
        <li>
          Ingredient 2
        </li>
        <li>
          Ingredient 3
        </li>
      </ul>
      <p
        class="sc-gsnTZi FuqXk"
        data-testid="recipe-line"
      >
        <span
          class="sc-dkzDqf IGaFb"
        >
          Total time:
        </span>
         40min
      </p>
    </section>
  </section>
</DocumentFragment>
`;
